src/hypothesis/internal/conjecture/engine.py

```diff
@@ -1379,9 +1379,12 @@ def remove_discarded(self):
         """Try removing all bytes marked as discarded."""
         if not self.shrink_target.discarded:
             return
-        attempt = bytearray(self.shrink_target.buffer)
-        for u, v in sorted(self.shrink_target.discarded, reverse=True):
-            del attempt[u:v]
+        include = [True] * len(self.shrink_target.buffer)
+        for u, v in self.shrink_target.discarded:
+            for i in hrange(u, v):
```

Comment:
This looks worst-case quadratic! I think we'd be better off reusing the machinery in [`_union_intervals`](https://github.com/HypothesisWorks/hypothesis-python/blob/611da0df747a87cdda9b4b5ece7aa59fed3e8a8b/src/hypothesis/internal/charmap.py#L109) (possibly extracted somewhere) to normalise the discarded intervals, and then sticking with the obvious pop-and-delete method.

Reply:
Fair point. I don't think we can easily reuse `_union_intervals` due to slightly different intervals (differences over whether the end points are included), but I realised that there's an easy solution - we just check if any discarded intervals were added when we finish an interval and remove them if the current interval is discarded!